src/Servers/Kestrel/Transport.Sockets/src/WebHostBuilderSocketExtensions.cs

```diff
@@ -25,7 +27,14 @@ public static IWebHostBuilder UseSockets(this IWebHostBuilder hostBuilder)
     {
         return hostBuilder.ConfigureServices(services =>
         {
-            services.AddSingleton<IConnectionListenerFactory, SocketTransportFactory>();
+            services.TryAddSingleton<IConnectionListenerFactory, SocketTransportFactory>();
+
+            services.TryAddSingleton<IMemoryPoolFactory<byte>, DefaultSimpleMemoryPoolFactory>();
+            services.AddOptions<SocketTransportOptions>().Configure((SocketTransportOptions options, IMemoryPoolFactory<byte> factory) =>
+            {
+                // Set the IMemoryPoolFactory from DI on SocketTransportOptions. Usually this should be the PinnedBlockMemoryPoolFactory from UseKestrelCore.
+                options.MemoryPoolFactory = factory;
```

Comment:
Given that we're changing the type of the MemoryPoolFactory from a Func to an IMemoryPoolFactory meaning we're breaking reflection anyway, why flow the factory through the options instead of constructor injecting it in SocketTransportFactory? Just to minimize churn? I guess this also ensures anyone constructing it themselves with an options instance resolved from DI gets the appropriate IMemoryPoolFactory.

I think it's fine to leave it, but I think it's our last chance not to flow the factory through options if we don't want to.

Reply:
It would mean new API (since the type is public) or using the internal impl trick and switching to that. We'd need to do it on NamedPipes as well.